File: leetcode_algorithm_stone/101.对称二叉树.py
#
# @lc app=leetcode.cn id=101 lang=python3
#
# [101] 对称二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import collections
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        queue = collections.deque([(root.left, root.right)])
        while queue:
            p, q = queue.popleft()
            if not p and not q:
                continue
            if not p or not q or (p.val != q.val):
                return False
            queue.append((p.left, q.right))
            queue.append((p.right, q.left))
        return True
    
    # # 递归方式
    # def isSymmetric(self, root: TreeNode) -> bool:
    #     def helper(p: TreeNode, q: TreeNode) -> bool:
    #         if not p and not q:
    #             return True
    #         if not p or not q:
    #             return False
    #         return p.val == q.val and helper(p.left, q.right) and helper(p.right, q.left)
    #     if not root:
    #         return True
    #     return helper(root.left, root.right)
# @lc code=end

File: leetcode_algorithm_stone/101.对称二叉树.cpp
/*
 * @lc app=leetcode.cn id=101 lang=cpp
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <iostream>
#include <queue>
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
        std::queue<TreeNode*> q;
        q.push(root->left);
        q.push(root->right);
        while (!q.empty()) {
            TreeNode* p1 = q.front(); q.pop();
            TreeNode* p2 = q.front(); q.pop();
            if (!p1 && !p2) continue;
            if (!p1 || !p2 || p1->val != p2->val) return false;
            q.push(p1->left);
            q.push(p2->right);
            q.push(p1->right);
            q.push(p2->left);
        }
        return true;
    }


    // bool isSymmetric(TreeNode* root) {
    //     if (!root) return false;
    //     return helper(root->left, root->right);
    // }

    // bool helper(TreeNode *p, TreeNode *q) {
    //     if (!p && !q) return true;
    //     if (!p || !q) return false;
    //     return (p->val == q->val) && helper(p->left, q->right) && helper(p->right, q->left);
    // }
};
// @lc code=end

